### DISCUSSION TITLE
`@ogma/nestjs-module`: How to test controllers/services?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [jmcdo29/ogma#1538] [Q&A] [answered]
In `app.services.spec.ts`, I defined `app` as follows (partially based on [`packages/nestjs-module/test/ogma.service.spec.ts`](https://github.com/jmcdo29/ogma/blob/main/packages/nestjs-module/test/ogma.service.spec.ts#L31-L54)):

```ts
const app = await Test.createTestingModule({
  providers: [
    ConfigService,
    OgmaService,
    GatewayService,
    {
      provide: 'OGMA_INSTANCE',
      useFactory: () => new Ogma({
        logLevel: 'ALL' as const,
        color: true,
      }),
    },
    {
      provide: 'OGMA_CONTEXT',
      useValue: 'Test_context',
    },
    {
      provide: 'OGMA_SERVICE:GatewayService',
      useValue: 'GatewayService',
    },
    {
      provide: 'OGMA_TRACE_METHOD_OPTION',
      useValue: 'fine',
    },
  ],
}).compile()
```

It works, but is it the correct way to mock `OgmaService`?

Thanks! :pray:

**Comment by jmcdo29** [comment, marked as answer]
If you use `OgmaService` directly you can just use a simple custom provider:

```ts
{
  provide: OgmaService,
  useValue: {
    log: jest.fn(),
    debug: jest.fn(),
    ...etc
  }
}
```

If you use `@OgmaLogger()` you can do the same as above but use `createProviderToken()` instead of `OgmaService`

**Reply by tukusejssirs** [reply]
> If you use `OgmaService` [or] `@OgmaLogger()`

Do you mean in `app.service.ts`? I use `@OgmaLogger()`. I don’t want to test if Ogma outputs the logs correctly; all I want is to make Jest and Nest happy that the module is correctly defined. :wink:

Therefore I should simply use `jest.fn()`. The following still complains that `Nest can't resolve dependencies of the GatewayService (ConfigService, ?). Please make sure that the argument OGMA_SERVICE:GatewayService at index [1] is available in the RootTestModule context.`

```ts
const app = await Test.createTestingModule({
  providers: [
    ConfigService,
    GatewayService,
    {
      provide: OgmaService,
      useValue: {
        log: jest.fn(),
        debug: jest.fn(),
      },
    },
    // {
    //    provide: 'OGMA_SERVICE:GatewayService',
    //    useValue: 'GatewayService',
    // }_
  ],
}).compile()
```

After I uncomment the commented out lines, all works fine.

Now, If I replace `provide: OgmaService` with `provide: createProviderToken('OgmaService')`, it still misses `OGMA_SERVICE:GatewayService`; is its definition necessary?

Thanks for such prompt response! :pray:

**Reply by jmcdo29** [reply]
`createProviderToken('GatewayService')`. It takes the place of `@OgmaLogger(GatewayService)` and ends up creating the proper injection token for you, that `OGMA_SERVICE:GatewayService`.

Think of the `createProviderToken` analogous to `getRepositoryToken()` and `@InjectRepository()` from `@nestjs/typeorm` or the similar from `@nestjs/mongoose` and `@nestjs/sequelize`

**Reply by tukusejssirs** [reply]
Oh, why did I thought I use `createProviderToken('OgmaService')`? :thinking:

Thanks! Now it works as expected. And sorry I asked you a question only marginally related to Ogma. :wink:

**Reply by jmcdo29** [reply]
> And sorry I asked you a question only marginally related to Ogma. wink

It's related to the testing of the package. Always helpful to have the extra info laying around
